src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,161 @@
+        {
+            var newDocument = await ComputeNewDocumentAsync(document, item, cancellationToken).ConfigureAwait(false);
+            var newText = await newDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            var changes = await newDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
+            var change = Utilities.Collapse(newText, changes.ToImmutableArray());
+
+            return CompletionChange.Create(change);
+        }
+
+        private async Task<Document> ComputeNewDocumentAsync(Document document, CompletionItem completionItem, CancellationToken cancellationToken)
+        {
+            var containingNamespace = TypeImportCompletionItem.GetContainingNamespace(completionItem);
+            Debug.Assert(containingNamespace != null);
+
+            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var root = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            // Complete type name.
+            var textWithTypeName = root.GetText(text.Encoding).Replace(completionItem.Span, completionItem.DisplayText);
+            var documentWithTypeName = document.WithText(textWithTypeName);
+
+            // Find added node so we can use it to decide where to insert using/imports.
+            var treeWithTypeName = await documentWithTypeName.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var rootWithTypeName = await treeWithTypeName.GetRootAsync(cancellationToken).ConfigureAwait(false);
+            var addedNode = rootWithTypeName.FindToken(completionItem.Span.Start, findInsideTrivia: true).Parent;
+
+            // Add required using/imports directive.                              
+            var addImportService = documentWithTypeName.GetLanguageService<IAddImportsService>();
+            var optionSet = await documentWithTypeName.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var placeSystemNamespaceFirst = optionSet.GetOption(GenerationOptions.PlaceSystemNamespaceFirst, documentWithTypeName.Project.Language);
+            var compilation = await documentWithTypeName.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            var importNode = CreateImport(documentWithTypeName, containingNamespace);
+
+            var rootWithImport = addImportService.AddImport(compilation, rootWithTypeName, addedNode, importNode, placeSystemNamespaceFirst);
+            var documentWithImport = documentWithTypeName.WithSyntaxRoot(rootWithImport);
+
+            // Format newly added nodes.
+            return await Formatter.FormatAsync(documentWithImport, Formatter.Annotation, cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
1. is this necessary?
2. this is very heavyweight. you're formatting the entire doc. 

If you really need to do this for some reason, better would be to put an annotation on the import, then go find it post-insert, then format that range.

Reply:
Hmm, I thought this is formatting nodes with annotation only (I passed in Formatter.Annotation)? Is it still expensive?

Reply:
i'm dumb.  ignore me.